table/block_based/block_based_table_reader.cc

```diff
@@ -805,23 +796,29 @@ Status BlockBasedTable::Open(
 
 Status BlockBasedTable::PrefetchTail(
     const ReadOptions& ro, RandomAccessFileReader* file, uint64_t file_size,
-    bool force_direct_prefetch, TailPrefetchStats* tail_prefetch_stats,
-    const bool prefetch_all, const bool preload_all,
-    std::unique_ptr<FilePrefetchBuffer>* prefetch_buffer, Statistics* stats) {
+    bool force_direct_prefetch, const bool prefetch_all, const bool preload_all,
+    std::unique_ptr<FilePrefetchBuffer>* prefetch_buffer, Statistics* stats,
+    uint64_t tail_start_offset, bool contain_no_data_block,
+    Logger* const logger) {
+  assert(tail_start_offset <= file_size);
+
   size_t tail_prefetch_size = 0;
-  if (tail_prefetch_stats != nullptr) {
-    // Multiple threads may get a 0 (no history) when running in parallel,
-    // but it will get cleared after the first of them finishes.
-    tail_prefetch_size = tail_prefetch_stats->GetSuggestedPrefetchSize();
-  }
-  if (tail_prefetch_size == 0) {
-    // Before read footer, readahead backwards to prefetch data. Do more
-    // readahead if we're going to read index/filter.
-    // TODO: This may incorrectly select small readahead in case partitioned
-    // index/filter is enabled and top-level partition pinning is enabled.
-    // That's because we need to issue readahead before we read the properties,
-    // at which point we don't yet know the index type.
-    tail_prefetch_size = prefetch_all || preload_all ? 512 * 1024 : 4 * 1024;
+  if (tail_start_offset != 0) {
+    tail_prefetch_size = file_size - tail_start_offset;
+  } else if (contain_no_data_block) {
+    tail_prefetch_size = file_size - tail_start_offset;
+    ROCKS_LOG_WARN(logger,
+                   "Tail prefetch size %zu is calculated based on tail start "
+                   "offset being 0 and the "
+                   "file does not contain any data blocks",
+                   tail_prefetch_size);
+  } else {
+    tail_prefetch_size = prefetch_all || preload_all
+                             ? static_cast<size_t>(4 * 1024 + 0.01 * file_size)
```

Comment:
@anand1976 @ajkr I welcome any better heuristics for the case where we don't have tail_prefetch_size e.g, backward compatibility. I used the current one simply because someone provided it in the task about this.

Reply:
I worry a little bit about smooth upgrade. When someone upgrades to this version, they'll initially get this crude heuristic rather than the old adaptive TailPrefetchStats. How hard would it be to leave TailPrefetchStats for a release or two as a fall-back when the new data is not available?

Reply:
Good point - it's not difficult to leave TailPrefetchStats for a couple more release as a fallback and I will fix that.

Reply:
Fixed